src/dune_engine/package.ml

```diff
@@ -63,10 +63,10 @@ module Id = struct
       ; dir : Path.Source.t
       }
 
-    let compare { dir; name } pkg =
-      match Name.compare pkg.name name with
-      | Eq -> Path.Source.compare dir pkg.dir
-      | s -> s
+    let compare { name; dir } pkg =
+      let open Ordering.O in
+      let= () = Name.compare name pkg.name in
```

Comment:
Note that here we used to compare package names in the wrong order. This was never noticed because the only place where comparison matters is converting a set to a list and then converting that list back to a different set.